### PR TITLE
asadiqbal08/sol 1824 - Complete README instructions for xblock-poll

### PR DESCRIPTION
**Pull request by asadiqbal08**
[SOL-1824](https://openedx.atlassian.net/browse/SOL-1824)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -274,6 +274,13 @@ i18n_tool generate
 Before running the `i18n_tool extract` or `i18n_tool generate` commands, appropriate settings for xblock-poll need to be added to `edx-platform/conf/locale/config.yaml`.
 You should have to mention `poll` as a third_party app and `poll.po` in `django.po` list under `generate_merge` section.
 
+You can move the files into xblock-poll repo via using unix `mv` command
+* 'cd <path to edx-platform repo>/'
+* 'mv conf/locale/en/LC_MESSAGES/poll.po <Path to Xblock_Poll repo>/poll/translations/en/LC_MESSAGES/text.po'
+* 'mv conf/locale/en/LC_MESSAGES/poll.mo <Path to Xblock_Poll repo>/poll/translations/eo/LC_MESSAGES/text.mo'
```
Comment (mattdrayer) [inline README.md]: do we want to 'mv' these files or 'cp' them?

Comment (asadiqbal08) [inline README.md]: I think, `cp` is the good choice.

**Comment by asadiqbal08** [comment]
@itsjeyd kindly review the change set !

**Comment by itsjeyd** [comment]
@asadiqbal08 Thanks for following up on the previous PR! I'll try and have a look at this later on today.

**Comment by itsjeyd** [comment]
@asadiqbal08 I had a look at the changes. Here's my feedback:

1. The README should mention the directory from which to run the `i18n_tool extract` and `i18n_tool generate` commands. Basically, change this part

   ```
   For each language, create a domain file named `text.po` and compile it: You can use edX's [i18n-tools](https://github.com/edx/i18n-tools) to extract translations for multiple languages as follows:
   ```

   to

   ```
   For each language, create a domain file named `text.po` and compile it. You can use edX's [i18n-tools](https://github.com/edx/i18n-tools) to extract translations for multiple languages into domain files by running the following command from the top-level directory of your edx-platform installation (`/edx/app/edxapp/edx-platform`):
   ```

2. Formatting is a little inconsistent for this paragraph:

   ```
   Before running the `i18n_tool extract` or `i18n_tool generate` commands, appropriate settings for xblock-poll need to be added to `edx-platform/conf/locale/config.yaml`. You should have to mention `poll` as a third_party app and `poll.po` in `django.po` list under `generate_merge` section.
   ```

   My suggestion would be to add an example and change it to:

   ```
   Before running `i18n_tool extract` or `i18n_tool generate` as described above, appropriate settings for xblock-poll need to be added to `edx-platform/conf/locale/config.yaml`. In particular, you'll need to add `poll` as a `third_party` app and list `poll.po` under `django.po` in the `generate_merge` section:

       ...
       third_party:
           - ...
           - poll
       ...
       generate_merge:
           django.po:
               - ...
               - poll.po
   ```

3. The rendered Markdown for the paragraph you added doesn't show the parts surrounded by angle brackets (`<` and `>`). You can fix that by changing it to:

   ```
   As a last step, copy the translation files into the xblock-poll repo by running the following commands from the top-level directory of your edx-platform installation (`/edx/app/edxapp/edx-platform`):

       cp conf/locale/en/LC_MESSAGES/poll.po /path/to/xblock-poll/poll/translations/en/LC_MESSAGES/text.po
       cp conf/locale/en/LC_MESSAGES/poll.mo /path/to/xblock-poll/poll/translations/en/LC_MESSAGES/text.mo

   You can repeat these commands for other languages as necessary. Be sure to replace `/path/to/xblock-poll` with the actual path to the xblock-poll repo.
   ```

---

I also tested the instructions again by following the steps for

* adding translation files for another language
* updating existing translation files

Here is what I found:

1. Generating .po/.mo files for another language (I tried `fr`) did not work: After running `i18n_tool extract` and `i18n_tool generate`, the `conf/locale/fr/LC_MESSAGES/` directory did not contain `poll.po` and `poll.mo` files, so I couldn't add them to the xblock-poll repo.

2. Changes that I made to one of the strings in `xblock-poll/poll/poll.py` did not get picked up when extracting translations. The new version of `text.po` for English still contained the old version of the string. Here is the (abbreviated) patch for the changes that I made:

   ```
   -            result['errors'].append(ugettext('Maximum Submissions missing or not an integer.'))
   +            result['errors'].append(ugettext('TEST EXTRACTION'))
   ```

   I would have expected the `text.po` file to contain an entry for `'TEST EXTRACTION'`, but that wasn't the case; it still contained the `'Maximum Submissions missing or not an integer.'` string I had removed from the code.

**Comment by asadiqbal08** [comment]
@itsjeyd I have updated the readme. 
It seems me, `i18n_tool` works for `en` and `eo` (generate dummy translations e.g. via `i18n_tool dummy`). The purpose of this tool is to extract/generate/compile `en` or `eo` translated files with strings in source codes. We can verify it from `i18n_tool` author / developers.  @mattdrayer  fyi here.

File: README.md [hunk 2]
```diff
@@ -271,15 +271,24 @@ Open each `text.po` domain file and, for each `msgid` string, add a correspondin
...
-* 'cp conf/locale/en/LC_MESSAGES/poll.mo <Path to Xblock_Poll repo>/poll/translations/eo/LC_MESSAGES/text.mo'
+...
+   third_party:
+       - ...
+       - poll
+   ...
+   generate_merge:
+       django.po:
+           - ...
+           - poll.po
```
Comment (itsjeyd) [inline README.md]: @asadiqbal08 This doesn't render correctly:

broken-markdown

Each of the lines in this example (L276-L284) needs to start with four spaces (currently it's only three).

File: README.md [hunk 3]
```diff
@@ -271,15 +271,24 @@ Open each `text.po` domain file and, for each `msgid` string, add a correspondin
...
 
-You can run the above command for `eo` as well.
+
+As a last step, copy the translation files into the xblock-poll repo by running the following commands from the top-level directory of your edx-platform installation (`/edx/app/edxapp/edx-platform`):
```
Comment (itsjeyd) [inline README.md]: @asadiqbal08 Nit: Insert an empty line after this one.

**Comment by itsjeyd** [comment]
@asadiqbal08 Thanks for the updates. There's just a couple of small formatting inconsistencies left to fix for the existing content.

> It seems me, `i18n_tool` works for `en` and `eo` (generate dummy translations e.g. via `i18n_tool dummy`). The purpose of this tool is to extract/generate/compile `en` or `eo` translated files with strings in source codes.

I'm not sure what you mean -- are you saying that I should be using `i18n_tool dummy` instead of `i18n_tool extract`/`i18n_tool generate` to extract and compile translations?

As a general comment/reminder, the goal is for someone who does not know anything about working with translations to be able to successfully create/update translations (for any language) by following the instructions in the README.

This means that we can't leave out any mandatory steps, and that descriptions of individual steps must be accurate (produce desired results). If something does not work yet (for example, if it is not possible to generate translations for languages other than `en` and `eo`), the README should mention this.

**Comment by asadiqbal08** [comment]
@itsjeyd I think, we mention it in readme that it is not possible to generate translations for languages other than `eo` that is dummy translated strings. in case of `en` tool extract the strings and place them in .po but this file does not contain translations. as you can see [HERE](https://github.com/edx/edx-platform/blob/master/conf/locale/en/LC_MESSAGES/django.po)
However, Can we conform from `i18n_tool` author ?

**Comment by mattdrayer** [comment]
@nedbat, can you provide some insight here?  We have a question regarding I18N string extractions using the i18n-tools utility.

**Comment by itsjeyd** [comment]
@asadiqbal08 I think there's been a misunderstanding: I know that it's not possible to automatically generate *actual translations* for languages other than `eo`. My [comment](https://github.com/open-craft/xblock-poll/pull/12#issuecomment-226256425) was about translation *files*: I was assuming that the `i18n_tool extract` command would automatically put translation files into `LC_MESSAGES` directories for each language listed in `config.yaml` (`conf/locale/en/LC_MESSAGES`, `conf/locale/fr/LC_MESSAGES`, `conf/locale/ar/LC_MESSAGES`, etc.).

But IIUC, this is not the case. Instead, the `i18n_tool extract` command produces a `poll.po` file for English (`conf/locale/en/LC_MESSAGES/poll.po`) that can be reused for other languages. So instead of doing

```sh
cp conf/locale/fr/LC_MESSAGES/poll.po /path/to/xblock-poll/poll/translations/fr/LC_MESSAGES/text.po
```

you would do

```sh
cp conf/locale/en/LC_MESSAGES/poll.po /path/to/xblock-poll/poll/translations/fr/LC_MESSAGES/text.po
```

to add a translation file for French to xblock-poll. Is that correct?

**Comment by asadiqbal08** [comment]
@itsjeyd Yes. 
What you think if use the `<lang_code>` instead of specific language as you mentioned above the `fr` ? 
like that : 
`cp conf/locale/en/LC_MESSAGES/poll.po /path/to/xblock-poll/poll/translations/<Lang_Code>/LC_MESSAGES/text.po`
Lang_Code can be `fr`, `ar`, `en` etc

**Comment by itsjeyd** [comment]
@asadiqbal08 Thanks for the info. We could change the command for copying translation files as you suggest.

However, I'm wondering if it is necessary to use `i18n_extract` at all if it doesn't generate translation files for languages other than English automatically: The xblock-poll repo ships with a `text.po` file for English (introduced in #10). This file could be used as a starting point for creating translations for other languages, right?

If that's the case, the workflow described in "Adding Translations for Other Languages" could be reduced to:

1. Add `poll/translations/<lang_code>/LC_MESSAGES/` directories for desired languages.
2. Copy `text.po` file from `poll/translations/en/LC_MESSAGES/` to each of the new directories.
3. Open each `text.po` domain file and, for each `msgid` string, add a corresponding `msgstr` translation.
4. Compile `text.po` files as described in previous section (run `msgfmt text.po -o text.mo` on it).

Is that correct?

**Comment by asadiqbal08** [comment]
@itsjeyd The purpose of `i18n_extract` is basically, extract the strings from sources files and put them in '.po' in edx-platform. It does not generate translations. I think, This command is useful.

**Comment by itsjeyd** [comment]
@asadiqbal08 Yes, I'm aware of what the `i18n_tool extract` command does. But based on my current understanding, I don't see a reason for using it to "Add Translations for Other Languages". Can you please answer the question from my previous comment? The question was: Do the following steps describe a complete workflow for "Adding Translations for Other Languages"?

1. Add `poll/translations/<lang_code>/LC_MESSAGES/` directories for desired languages.
2. Copy `text.po` file from `poll/translations/en/LC_MESSAGES/` to each of the new directories.
3. Open each `text.po` domain file and, for each `msgid` string, add a corresponding `msgstr` translation.
4. Compile `text.po` files as described in previous section (run `msgfmt text.po -o text.mo` on it).

Note that I am not asking you to update the README just yet; I am just looking for a straightforward answer to my question.

**Comment by asadiqbal08** [comment]
Question: The xblock-poll repo ships with a text.po file for English (introduced in #10). This file could be used as a starting point for creating translations for other languages, right?

Ans: Yes

The steps that you have mentioned above. Looks good to me

**Comment by nedbat** [comment]
Hi, sorry I'm just catching up to this.  I have a few concerns about these instructions.

First, why are we telling people to modify the edx-platform files in order to get translations for an XBlock? I thought we had a workflow that let us work with the XBlock independently of the edx-platform repo.  Are we meant to check-in that change to edx-platform? If so, then edx-platform will have references to all the XBlocks? And if we aren't meant to check it in, then it's a step we have to repeat and clean up each time? That's not a reasonable plan.  The i18n-tools repo is separate from edx-platform specifically so it could be used in other repos without touching edx-platform at all.

Second, the instructions seem odd to me.  They say, "For each language, create a domain file named `text.po` and compile it," but don't explain how to create it, or how to compile it.  Then it says the developer needs to open a .po file and edit it by hand to add empty translations for all the strings?  I don't understand how the .po file got made without the msgstr entries in the first place.

Lastly, these are not instructions particular to the poll XBlock. Whatever steps we settle on, the instructions should be in a more central place so that any XBlock author can find them.

**Comment by itsjeyd** [comment]
@asadiqbal08 OK, thanks.

Thanks for the feedback, @nedbat. We are in the process of reworking the instructions in the README. Due to time constraints, the PR that originally introduced them (#10) had to be merged before the instructions were finalized; hence this follow-up PR and [associated story](https://openedx.atlassian.net/browse/SOL-1824). Based on the latest information from @asadiqbal08 we should be able to remove any instructions about modifying edx-platform files from the README.

> Lastly, these are not instructions particular to the poll XBlock. Whatever steps we settle on, the instructions should be in a more central place so that any XBlock author can find them.

The idea was for the README to describe how to work with translation files in the context of the xblock-poll repo (so that a developer reading the instructions wouldn't necessarily need to consult additional resources), and to link to the [Internationalization Support](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support) section of the XBlock tutorial for additional info/context.

**Comment by nedbat** [comment]
@itsjeyd I really appreciate all the work that has gone into the XBlock i18n work, it's been a gap for a long time, and lots of people will make use of it.  I just want to make sure that the information is useful to the most people.  These instructions will be the same for any XBlock, right?  It will really be best for information common to all XBlocks to be in a common location, so that each XBlock doesn't have to duplicate it.  I understand wanting to keep all the info about this XBlock here, but developers working with XBlocks will need to understand a number of concepts common to all blocks, and we won't be duplicating all the information here.

One possibility is to finish this work here, and then move it to a common location, but that seems like extra work.  Can we shift slightly, and put this information in a common location now?

**Comment by itsjeyd** [comment]
@nedbat Yes, I think the instructions should be the same across XBlocks, at least for strings defined in Python code. This XBlock includes some Handlebars templates ([1](https://github.com/open-craft/xblock-poll/blob/master/poll/public/handlebars/poll_results.handlebars#L31), [2](https://github.com/open-craft/xblock-poll/blob/master/poll/public/handlebars/poll_studio.handlebars#L6), [3](https://github.com/open-craft/xblock-poll/blob/master/poll/public/handlebars/survey_results.handlebars#L30)) with custom [i18n helpers](https://github.com/open-craft/xblock-poll/blob/master/poll/public/js/poll.js#L20) that will require special handling when it comes to extracting strings. So these might require instructions that are specific to xblock-poll. But AFAIK, the work for automating the process of extracting strings from these files is still outstanding (@asadiqbal08 @mattdrayer please correct me if I'm wrong), so we were not going to add information about that to the README for now.

I would be fine with putting common information in a common location, and changing the README for this repo to point readers there. @bradenmacdonald @mattdrayer @asadiqbal08 What do you guys think?

**Comment by bradenmacdonald** [comment]
+1, don't repeat yourself. A common location makes sense.

There is some existing documentation in the XBlock tutorial (although IMHO, authoritative docs should not be located in a tutorial): http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support

And I wrote up instructions for testing XBlocks with dummy translations in my XBlock presentation at the conference: https://docs.google.com/presentation/d/1D5zB6yWfv-r7egMAfKpslf8JOM3y0C1DmldaPWIyf5g/edit#slide=id.g14db027f83_0_44

**Comment by bradenmacdonald** [comment]
BTW I want to second Ned's comment, that there should be a way to use the i18n tools to extract strings and generate dummy translations without needing to modify or involve edx-platform in any way. If I recall, the `extract` and `dummy` commands are currently only working in the context of edx-platform, and also require modifying the edx-platform i18n config.

**Comment by asadiqbal08** [comment]
@itsjeyd for that particular story, Can you refine the text so I can update it ? we can make separate stories for future sprints.

**Comment by itsjeyd** [comment]
@bradenmacdonald @nedbat Do you have any preferences about where to put the instructions? One option would be to extend the [Internationalization Support](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support) section of the XBlock tutorial. In addition to the information that it currently provides, the following topics might be worth covering (in no particular order):

* As a developer, what do I need to do to view an XBlock in a different language? Is the workflow different when using dummy vs. real translations?
* As a developer, how do I update translation files after modifying strings in the source code? Do I have to edit translation files manually, or is there a way to (partially) automate the process?
* Transifex integration: How can I outsource string translation for my XBlock?

I also think it would be nice to have an example available that demonstrates workflows for specific use cases using concrete commands: The XBlock tutorial lists the tools that can be used to extract strings and compile translation files, but it doesn't show how to use them (except for the `i18n_tool extract` and `i18n_tool generate` commands, but IIUC these require creating an "alternative configuration file" if they are to be used independently of edx-platform).

@bradenmacdonald @mattdrayer @asadiqbal08 Then there's the question of scope -- writing instructions that aren't specific to xblock-poll (and cover some or all of the points above) might be out-of-scope for [SOL-1824](https://openedx.atlassian.net/browse/SOL-1824)?

@asadiqbal08 We'll need to clarify the points above before making further changes to the README. Once that happens, we should be able to shorten the text to one or two sentences pointing the reader to relevant external resources.

**Comment by mattdrayer** [comment]
I would like to get to a decision on how to proceed with this work, so we can finish the story.  We are halfway through our sprint, so there's plenty of time to add documentation wherever we would like for it to live.  I agree that it makes sense to expand the Internationalization Support section of the XBlock documentation at RTD.

**Comment by nedbat** [comment]
Seems like the XBlock tutorial is a fine place to add this information.

**Comment by itsjeyd** [comment]
@mattdrayer @nedbat OK thanks, let's go ahead and shift the focus of the story to the XBlock tutorial then. From my perspective, the next steps would be:

1. Finish this PR: Remove the subsections of the "Working with Translations" section ("Viewing the XBlock in a Different Language", "Adding Translations for Other Languages", "Updating Existing Translation Files") from the README and reduce the remaining text to something like:

   > For information about working with translations, see the [Internationlization Support](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support) section of the [Open edX XBlock Tutorial](https://xblock-tutorial.readthedocs.io/en/latest/).

2. Open a new PR against the edx-documentation repo that adds information about some or all of the topics listed above.

If there are no objections, please go ahead with the steps listed above @asadiqbal08.

CC @bradenmacdonald

**Comment by asadiqbal08** [comment]
@itsjeyd I have shorted the readme, take a look. If we are good with this particular story than we can merge it. 

##### 2 - Open a new PR against the edx-documentation repo that adds information about some or all of the topics listed above.
Are you saying that we should add the "Viewing the XBlock in a Different Language", "Adding Translations for Other Languages", "Updating Existing Translation Files" sections in edx-documentaion repo ? If yes, then in which file(s) should I have to add this information in edx-documentation repo ? 

Thanks.

File: README.md [hunk 4]
```diff
@@ -240,40 +240,4 @@ names in the django settings using the `XBLOCK_POLL_EXTRA_VIEW_GROUPS` setting,
...
-```sh
-i18n_tool generate
-```
-Before running the `i18n_tool extract` or `i18n_tool generate` commands, appropriate settings for xblock-poll need to be added to `edx-platform/conf/locale/config.yaml`.
-You should have to mention `poll` as a third_party app and `poll.po` in `django.po` list under `generate_merge` section.
-
-### Updating Existing Translation Files
-
-After changing existing strings in the source code (or adding new ones), you'll need to update the translation files. To do this, extract and compile the translations as described in the previous section. (There is no need to create additional directories if you are not adding translations for additional languages.)
+For information about working with translations, see the [Internationalization](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support) Support section of the [Open edX XBlock Tutorial](https://xblock-tutorial.readthedocs.io/en/latest/).
```
Comment (itsjeyd) [inline README.md]: @asadiqbal08 Can you make "Support" part of the link please?

Here's the Markdown:

```
For information about working with translations, see the [Internationalization Support](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support) section of the [Open edX XBlock Tutorial](https://xblock-tutorial.readthedocs.io/en/latest/).
```

Here is what the rendered version should look like:

For information about working with translations, see the [Internationalization Support](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support) section of the [Open edX XBlock Tutorial](https://xblock-tutorial.readthedocs.io/en/latest/).

**Comment by asadiqbal08** [comment]
done

**Comment by itsjeyd** [comment]
@asadiqbal08 OK thanks. Please squash your commits. :+1: once that's done and the build is green.

> Are you saying that we should add the "Viewing the XBlock in a Different Language", "Adding Translations for Other Languages", "Updating Existing Translation Files" sections in edx-documentaion repo ?

No, transferring the information from these sections without modification won't work I think. (The info from these sections partially overlaps with the info from the tutorial. And as Ned mentioned in [this comment](https://github.com/open-craft/xblock-poll/pull/12#issuecomment-227712408), instructions for working with translations should not involve edx-platform.)

I think it would be useful if the XBlock tutorial contained information about the topics I listed in [this comment](https://github.com/open-craft/xblock-poll/pull/12#issuecomment-228101267). I'll repeat them here:

> * As a developer, what do I need to do to view an XBlock in a different language? Is the workflow different when using dummy vs. real translations?
> * As a developer, how do I update translation files after modifying strings in the source code? Do I have to edit translation files manually, or is there a way to (partially) automate the process?
> * Transifex integration: How can I outsource string translation for my XBlock?
> 
> I also think it would be nice to have an example available that demonstrates workflows for specific use cases using concrete commands: The XBlock tutorial lists the tools that can be used to extract strings and compile translation files, but it doesn't show how to use them (except for the `i18n_tool extract` and `i18n_tool generate` commands, but IIUC these require creating an "alternative configuration file" if they are to be used independently of edx-platform).

Based on the current structure of the [Internationalization Support](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support) section, integrating this info will likely involve modifying the existing text as well as adding new sections.

I can provide some feedback once you've put together a first version of the changes. But note that you'll need to get an upstream review from someone on the documentation team for the new PR. (OpenCraft doesn't own the edx-documentation repo so we can't provide final thumbs.) 

> If yes, then in which file(s) should I have to add this information in edx-documentation repo ?

You'll have to modify this file: [edx_lms.rst](https://github.com/edx/edx-documentation/blob/master/en_us/xblock-tutorial/source/edx_platform/edx_lms.rst)

> we can create a separate story to track this. 

Up to @mattdrayer (but based on his latest [comment](https://github.com/open-craft/xblock-poll/pull/12#issuecomment-228391133) I think he wants to handle this as part of [SOL-1824](https://openedx.atlassian.net/browse/SOL-1824)).

**Comment by mattdrayer** [comment]
Merging this PR now -- we'll open a new edx-documentation PR for the corresponding changes there.  Assuming we can finish this work by the end of our current sprint, I am fine keeping track of everything in the SOL-1824 ticket.

**Comment by mattdrayer** [comment]
Oh, I can't merge -- @itsjeyd looks like it's on you :)

**Comment by itsjeyd** [comment]
@mattdrayer OK, I'll take care of merging the PR :)

@asadiqbal08 Please update the commit message. The current text ("added command instructions to copy .po/.mo files") no longer describes the changes. You could use something like this:

> Remove WIP documentation for working with translations.
> 
> As mentioned on https://github.com/open-craft/xblock-poll/pull/12, instructions for working with translations are (largely) the same across XBlocks, so it makes more sense to add them to the "Internationalization Support" section of the XBlock tutorial.

**Pull request merged**